### ISSUE TITLE
Delayed but sustained FPS drop while moving the camera view in Valheim

### ISSUE THREAD

**Issue by taleden** [BepInEx/BepInEx#197]
**Describe the bug**
When first launching, Valheim+BepInEx plays normally and loads any installed plugins. After an indeterminate length of play time (within ~1 hour), suddenly FPS starts dropping dramatically whenever the camera is moved. While stationary, FPS is normal (~25 in base, ~60 out, on my system), but simply jiggling the camera around slightly causes FPS to drop significantly (~8 in base, ~25 out). The UI is affected as well, so sound effects of clicking UI buttons are noticeably delayed after the actual click. The problem lasts until the game is exited and restarted, and then it runs well again for awhile until the issue reoccurs.

**To Reproduce**
Play Valheim for awhile with BepInEx loaded, possibly only on Linux (or I imagine this would have been noticed already).

**Expected behavior**
Performance should not be so severely impacted by simply moving the camera around. That it seems linked to the camera and not the player's position in the world makes me wonder if some sight-line or visibility function is to blame, since presumably that would be called every frame while the camera moves, but rarely or never while the viewport remains static.

**Screenshots and logs**
These show FPS of roughly the same scene, with the camera still and in motion, before and after the bug occurs. Motion blur was enabled just to clearly show which image has the camera in motion, but the bug persists with motion blur disabled. Images scaled down from 2560x1440 to save file size since the relevant data is still legible.
![before-still](https://user-images.githubusercontent.com/23055012/110025303-c857e500-7cf4-11eb-8d10-5fac1d655308.png)
![before-moving](https://user-images.githubusercontent.com/23055012/110025312-cbeb6c00-7cf4-11eb-9cc6-53be170f3310.png)
![after-still](https://user-images.githubusercontent.com/23055012/110025321-d0178980-7cf4-11eb-9c30-8b9eed995af5.png)
![after-moving](https://user-images.githubusercontent.com/23055012/110025330-d1e14d00-7cf4-11eb-9a45-fc06bf89e885.png)

**Desktop (please complete the following information):**
 - Valheim 0.147.3 on Linux Mint 20.1
 - BepInEx version 5.4.8
 - AMD Ryzen 7 1700 8-core x64 CPU
 - Nvidia GeForce GTX 1070 GPU
 - Nvidia 450.102.04 Linux-x86_64 GPU driver

**Additional context**
I also tried removing all BepInEx plugins, but the issue occurs even with BepInEx loaded and nothing else.

**Comment by ghorsington** [comment]
Greetings!

This does indeed seem like a peculiar issue. And it indeed does not seem easy to reproduce since it appears to take a long time before the issue is noticeable.

First things that comes to mind to check is RAM usage. Have you tried monitoring RAM usage for the game process over time while playing the game without BepInEx vs. with BepInEx? One obvious potential place for such lags could be a memory leak.
Could you possibly try monitoring how RAM changes over time in both cases (with and without bepin)? This would also work as a test whether the same lag appears in vanilla game.

I'm not 100% sure if Mint has some GUI process monitor by default, but I'd assume at least `top` or `htop` is available by default.

**Comment by taleden** [comment]
Good idea, but no leads; here's the resource monitor before the bug, taken while motionless after a typical session of running around base building stuff, followed by the monitor after the bug has triggered, taken while jiggling the camera to kill FPS:
![before-system](https://user-images.githubusercontent.com/23055012/110031743-e9243880-7cfc-11eb-9317-8422acb9f9d9.png)
![after-system](https://user-images.githubusercontent.com/23055012/110031765-ee818300-7cfc-11eb-9069-6d483b637e91.png)
Memory usage is constant (Valheim accounts for ~2GB of that), maybe a few CPU peaks in the after shot but no cores pegged or anything.

**Comment by taleden** [comment]
One more observation: since putting together this report I've watched the timing more carefully, and the last two times it happened almost exactly 35 minutes after starting. Dunno what that could correspond to, but I'll note if that pattern holds.

**Comment by taleden** [comment]
Just a shot in the dark, but do BepInEx's hooks into Unity functions allow for any kind of runtime call stack profiling? Would it be feasible to build a debugging plugin that just gathers stats on how long various main functions are taking to run, or how often? Or maybe such a thing already exists for Unity engine games in general? That could show us where to look.

**Comment by taleden** [comment]
Third time in a row, right in the 35-40 minute range. All I can think of so far is that Valheim now autosaves every 20 minutes, so 40 would be the second of the session; I'll keep an eye on my save files next time and see if the character file timestamp corresponds to when the lag starts.

**Comment by marcinstobba** [comment]
i have the same problem but without bepinex, and it dosnt happen on opengl, try to start the game from steam library, and check opengl

**Comment by taleden** [comment]
I did some more testing and resolved the problem for now, although I still don't know its root cause.

Based on @marcinstobba 's comment I tried running via OpenGL on linux, and while overall performance was noticeably worse, it did not suffer from this delayed lag problem (after ~2 hours).

I also tried again running without the BepInEx preloader, and that also had no issue (after 1.5 hours).

Just as a sanity check, I then ran again with Vulkan and the preloader and confirmed that the problem happened again (several times) in the 35-45 minute range after game start.

So next I tried booting back into Windows and ran using Vulkan on Windows and BepInEx (via winhttp.dll) on the same hardware, with no issue.

Then I happened across [this thread](https://www.reddit.com/r/linux_gaming/comments/lmslz9/best_settings_to_optimize_valheim/) which mentioned the Steam overlay. I had indeed disabled the overlay for Valheim when first trying to install BepInEx because I noticed in the console that it complained about invalid 32bit preloads, so I tried re-enabling the overlay and sure enough, the issue disappeared.

So it seems to be some combination of linux, Vulkan, BepInEx preload, but no Steam Overlay preload, which causes the problem.

One thing I haven't tried yet is testing other keyboard/mouse input like zooming the minimap, as mentioned in the thread, to see if that causes the same stutter as moving the camera. So it may actually be an input processing problem and not anything to do with rendering, and that in turn makes me wonder if it's something to do with BepInEx pulling in the un-stripped Unity libraries; maybe there's some incompatibility in an input handling library?

**Comment by ghorsington** [comment]
That would be quite peculiar since unstripped libraries included with the [official BepInExPack for Valheim](https://valheim.thunderstore.io/package/denikson/BepInExPack_Valheim/) are taken directly from Unity Editor 2019.4.20 so they are the ones they are truly supposed to be.

Are you able to use the overlay with BepInEx? It could be that those messages are harmless and just some sort of side-effect of BepInEx loading.

Unfortunately, there is no viable way to test theory of unstripped libraries since we would need original unstripped libs that are shipped with the game. Those would require the game devs to disable unstripping in the first place.

You mentioned you resolved the problem, however I'm not quite sure I understand your current situation: are you able to use BepInEx and not have the lag? Or do you need to have BepInEx disabled?

**Comment by taleden** [comment]
Yes, BepInEx works for long sessions with no issue after I re-enabled the Steam Overlay for Valheim. I'll test again today, but so far I've also had no issues in long sessions with the Steam Overlay disabled as long as BepInEx was also not loaded. So it's something about loading BepInEx but not the Overlay that does it; I have no idea what that would be, the unstripped libraries were just a thought based on others' reports that it's related to keyboard/mouse input handling (but strangely, not gamepads).

**Comment by ghorsington** [comment]
I've actually had a person report an issue with using a gamepad (specifically an Xbox One controller) with BepInEx where it would not get detected on game start and upon reconnecting it the trigger inputs got swapped. It could be somehow related. However, we won't know for sure until the developers stop stripping their assemblies.

This is quite peculiar that the overlay somehow fixes the issue. It feels as if when the overlay is enabled, something is not run which in turn fixes the lag. This indeed would warrant more testing.

**Comment by taleden** [comment]
Turns out this has nothing to do with BepInEx; sorry for the false alarm!

During all my testing, because I had configured Steam to launch Valheim using the BepInEx preloader shell script, I was doing my non-BepInEx testing by just directly running the game executable ("valheim.x86_64"), and that never had a problem. I assumed that since the game has a linux-native binary with Steam integration, and since the Steam client would detect that it was running when I started it this way, that this was effectively no different than running it via Steam with no special launch options. But that 's apparently not the case.

When I tried resetting the Steam launch options to run without BepInEx, but still with the Overlay disabled, then launching via Steam continues to exhibit this problem, while launching the game binary directly has no problem. So the factors that produce the issue are not linux + Vulkan + yes-BepInEx + no-Overlay, it's Steam-launched + no-Overlay; direct-launched + no-Overlay is fine, and Steam-launched + yes-Overlay is also fine.

So ultimately it must be a Steam issue; I'll see if I can figure out where to report it to them, and I'll close this ticket. Thanks for the suggestions, and sorry to bother!
